### DISCUSSION TITLE
Latency issue for token generation

### DISCUSSION THREAD

**Discussion by anup-krai** [Kong/kong#9185] [Help]
We are seeing high latency for oauth token generation calls. P90 is around 150ms. When we check from Postgres side the latency for insert statements was less than 1ms.
We checked the CPU and Memory utilisation for Kong pods and those were also normal. Is there any reason for this high latency for this call?
We have checked from network and Postgres side and latency seems to be very less.
We also noticed below few extra queries which are getting executed continuously at higher rate on Postgres side -
```
SELECT "id", EXTRACT($1 FROM "created_at" AT TIME ZONE $2) AS "created_at", "name", "algorithm", "hash_on", "hash_fallback", "hash_on_header", "hash_fallback_header", "hash_on_cookie", "hash_on_cookie_path", "slots", "healthchecks", "tags", "host_header", "client_certificate_id", "ws_id" FROM "upstreams" WHERE ("id") = ($3) AND ("ws_id" = $4) LIMIT $5
```
```
SELECT "id", EXTRACT($1 FROM "created_at" AT TIME ZONE $2) AS "created_at", "name", "algorithm", "hash_on", "hash_fallback", "hash_on_header", "hash_fallback_header", "hash_on_cookie", "hash_on_cookie_path", "slots", "healthchecks", "tags", "host_header", "client_certificate_id", "ws_id" FROM "upstreams" WHERE ("id") = ($3) LIMIT $4
```
```
INSERT INTO "upstreams" ("id", "created_at", "name", "algorithm", "hash_on", "hash_fallback", "hash_on_header", "hash_fallback_header", "hash_on_cookie", "hash_on_cookie_path", "slots", "healthchecks", "tags", "host_header", "client_certificate_id", "ws_id") VALUES ($1, TO_TIMESTAMP($2) AT TIME ZONE $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, ARRAY[$14], $15, $16, $17) ON CONFLICT ("id") DO UPDATE SET "created_at" = EXCLUDED."created_at", "name" = EXCLUDED."name", "algorithm" = EXCLUDED."algorithm", "hash_on" = EXCLUDED."hash_on", "hash_fallback" = EXCLUDED."hash_fallback", "hash_on_header" = EXCLUDED."hash_on_header", "hash_fallback_header" = EXCLUDED."hash_fallback_header", "hash_on_cookie" = EXCLUDED."hash_on_cookie", "hash_on_cookie_path" = EXCLUDED."hash_on_cookie_path", "slots" = EXCLUDED."slots", "healthchecks" = EXCLUDED."healthchecks", "tags" = EXCLUDED."tags", "host_header" = EXCLUDED."host_header", "client_certificate_id" = EXCLUDED."client_certificate_id", "ws_id" = EXCLUDED."ws_id" RETURNING "id", EXTRACT($18 FROM "created_at" AT TIME ZONE $19) AS "created_at", "name", "algorithm", "hash_on", "hash_fallback", "hash_on_header", "hash_fallback_header", "hash_on_cookie", "hash_on_cookie_path", "slots", "healthchecks", "tags", "host_header", "client_certificate_id", "ws_id"
```
```
INSERT INTO "services" ("id", "created_at", "updated_at", "name", "retries", "protocol", "host", "port", "path", "connect_timeout", "write_timeout", "read_timeout", "tags", "client_certificate_id", "tls_verify", "tls_verify_depth", "ca_certificates", "enabled", "ws_id") VALUES ($1, TO_TIMESTAMP($2) AT TIME ZONE $3, TO_TIMESTAMP($4) AT TIME ZONE $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, ARRAY[$15], $16, $17, $18, $19, $20, $21) ON CONFLICT ("id") DO UPDATE SET "created_at" = EXCLUDED."created_at", "updated_at" = EXCLUDED."updated_at", "name" = EXCLUDED."name", "retries" = EXCLUDED."retries", "protocol" = EXCLUDED."protocol", "host" = EXCLUDED."host", "port" = EXCLUDED."port", "path" = EXCLUDED."path", "connect_timeout" = EXCLUDED."connect_timeout", "write_timeout" = EXCLUDED."write_timeout", "read_timeout" = EXCLUDED."read_timeout", "tags" = EXCLUDED."tags", "client_certificate_id" = EXCLUDED."client_certificate_id", "tls_verify" = EXCLUDED."tls_verify", "tls_verify_depth" = EXCLUDED."tls_verify_depth", "ca_certificates" = EXCLUDED."ca_certificates", "enabled" = EXCLUDED."enabled", "ws_id" = EXCLUDED."ws_id" RETURNING "id", EXTRACT($22 FROM "created_at" AT TIME ZONE $23) AS "created_at", EXTRACT($24 FROM "updated_at" AT TIME ZONE $25) AS "updated_at", "name", "retries", "protocol", "host", "port", "path", "connect_timeout", "write_timeout", "read_timeout", "tags", "client_certificate_id", "tls_verify", "tls_verify_depth", "ca_certificates", "enabled", "ws_id"
```
Will these queries impact the oauth call latency?
Also we see sometimes KIC container CPU utilisation is going high during load. Can KIC impact the latency of API calls?
Kong version - 2.7.1
KIC - 2.2.1

**Comment by kikito** [comment]
Hello, please add more information about your system.

* It looks like you are using traditional gateway mode, can you confirm this? (e.g. you are not using hybrid mode)
* What is your approximate number of Routes, Customers, Services, Plugins, and Upstreams? Do you have active healthchecks in all  your Upstreams?
* How often do you add/remove Routes or Plugins?
* Which plugins are you using? In particular, any custom plugins?

**Reply by anup-krai** [reply]
Hi @kikito ,
Below are the details -
1) Yes are using traditional Kong gateway deployed in K8s environment with Kong ingress controller and Postgres DB.
2) Routes, services, upstreams are around 200, Consumers around 20, Plugins around 670. No health checks are configured for these services.
3) We add routes, plugins once is a day or two.
4) Plugins used for this are prometheus, requestSizing, cors, rateLimiting and one plugin for logging.

Also for testing this issue we have added few lines to capture the Postgres and network latency for oauth token generation calls [here](https://github.com/Kong/kong/blob/master/kong/plugins/oauth2/access.lua#L126). Below are the extra lines added to capture this -
```
  local timeBeforeDBCall = socket.gettime()
    token, err = kong.db.oauth2_tokens:insert({
      service = service_id and { id = service_id } or nil,
      credential = { id = credential.id },
      authenticated_userid = authenticated_userid,
      expires_in = token_expiration,
      refresh_token = refresh_token,
      scope = scope
    }, {
      -- Access tokens (and their associated refresh token) are being
      -- permanently deleted after 'refresh_token_ttl' seconds
      ttl = token_expiration > 0 and refresh_token_ttl or nil
    })
    local timeAfterDBCall = socket.gettime()
    local timeTaken = (timeAfterDBCall - timeBeforeDBCall) *1000
    kong.log.err("Total time taken for token insert in DB  : "..timeTaken)
    local set_serialize_value = kong.log.set_serialize_value
    set_serialize_value("oauth2_db_latency", timeTaken)
  end
```

We see Kong is adding ~10 ms to ~15 ms extra in response time. And this latency time increases as the load increases.

Below are few of the transactions time captured -
```
Total time taken for token insert in DB  : 6.3519477844238
```
For this call `X-Kong-Response-Latency` value in response is 16
```
 Total time taken for token insert in DB  : 10.658025741577
````
For this call `X-Kong-Response-Latency` value in response is 20
```
Total time taken for token insert in DB  : 41.758060455322
X-Kong-Response-Latency : 52
```
We have another instance of Kong gateway deployed in a different network segment with the same configuration where we see Kong is just adding ~1ms to ~2ms extra in token generation latency. And here count of routes, services and consumers are far more.

**Reply by kikito** [reply]
The call to `kong.db.oauth2_tokens:insert` that you measured does a direct write to the database.

Is it possible that the problem you are experimenting is not in Kong itself, but in the database, or in the connection between Kong and the database? Given that you have another Kong instance that isn't suffering from this problem, that could be a cause.
